### ISSUE TITLE
Support HTTP/S and Socks4/5 proxy connection

### ISSUE THREAD

**Issue by cimuxy** [Zelak312/flaresolverr-mitm-proxy#2]
How can I configure a SOCKS proxy to send the outbound requests via it ?

Ex . If I want to configure `flaresolverr-mitm-proxy:8080` as an upstream HTTP proxy server then the requests to this upstream proxy I want to make them to be sent via the SOCKS proxy I configured.

**Comment by Zelak312** [comment]
Hi, if I understand correctly you want the requests to go from
your proxy -> `flaresolverr-mitm-proxy:8080` -> internet

The setup required for configuring the flaresolverr proxy to be used as an upstream HTTP proxy server will depend on the software you use for your proxy server that you want to configure.

If you could give more information about the proxy server you use, I may be able to help you with the configuration needed to achieve this if the proxy server supports it

**Comment by cimuxy** [comment]
Hi, I mean that I want the request I send to go like this
The interceptor: `flaresolverr-mitm-proxy:8080` -> My Proxy: `user:password@global.rotating.proxyrack.net:10000` -> internet

the same as what BurpSuite do when I configure a socks proxy on it :
![Screenshot 2023-11-15 002731](https://github.com/Zelak312/flaresolverr-mitm-proxy/assets/93758098/ecf0d62e-820e-45b3-a3d0-a2334dd63881)

This will avoid us the detection when scraping a website or send POST data from the same IP address, Flaresolverr supports proxy connection already but we can't use it because we've configured `flaresolverr-mitm-proxy:8080` as a proxy already

for example If you have used BurpSuite, we configure the browser to use the BurpProxy in `127.0.0.1:8080` so in the browser we can't use another proxy because we used the BurpSuite's one, So we configure the proxy that we want to use in the BurpSuite it self 

I think the last example is identical to our case

My OS : Windows10

**Comment by Zelak312** [comment]
I see now.
From looking around at mitmproxy (what is used for flaresolverr-mitm-proxy). it doesn't seem possible to add a upstream SOCKS proxy as this issue says mitmproxy/mitmproxy#211 only http proxies are supported for now

There may be a simple solution to this problem.
By reading the comments on the mitmproxy issue, some people said to use proxychains to make this possible
If you are using the docker container, I can update it to add a way to configure upstream proxies http and socks by using proxychains.

If you manually are running mitmdump without docker, you will have to install proxychains and configure it to work as needed. I will mess around with proxychains and see if I can get what you want working and if you are not using docker, I can post an example of it if it works

**Comment by cimuxy** [comment]
Thank you very much, 
I think that I found a temporarily solution that to use HTTP proxy 
here is it : 
`mitmdump -s mitm.py --upstream http://username:password@my-upstream-proxy:port`
I hope to add this to README.md because due to the usage of flaresolverr-mitm-proxy we can't change our IP to avoid the CloudFlare detection, and this will lead to IP block.

Also I will try a tool like tun2socks or proxifier to achieve the SOCKS connection, I think the request will be like:
`flaresolverr-mitm-proxy:8080 -> http://127.0.0.1:tun2socks-proxifier-port -> http://username:password@my-upstream-proxy:port -> internet`

**Comment by eliasblume** [comment]
Hello, is there any good known setup on how to support socks5 ? I read the comments, but tbh i didn't get it.

**Comment by Zelak312** [comment]
I wrote about it a bit more on the [read me](https://github.com/Zelak312/flaresolverr-mitm-proxy?tab=readme-ov-file#using-an-upstream-proxy)

The current problem is that mitmproxy itself (what is used) doesn't support socks proxies as per the issue from their repo that I posted in this thread and also in the read me.

Now it's not impossible to make it work but it will require you to do more depending on how you use it and the OS you use.

If you looked at the issue I linked from the mitmproxy's repo, people talked about multiple ways to solve this. I seen
- Proxychains
- Gost
- Redsocks

If you are using the docker container, for now, it doesn't come with a build-in way to fix this issue but it could. I do plan to add it if it's requested.
In the container I plan to either use Gost or Proxychains, (in this order if one doesn't work with mitmproxy).

If you don't use the container then you will have to setup Proxychains, Gost or Redsocks yourself.
DISCLAIMER: Proxychains doesn't work on windows and I have no idea if they will work with mitmproxy.
I don't have any setup for any of them at the moment.

Some solutions may not also not work on Windows, like Proxychains I believe.
Here is something I found about using Gost https://github.com/mitmproxy/mitmproxy/issues/211#issuecomment-997330809
